app/src/main/java/org/jellyfin/mobile/ui/screens/connect/ServerSelection.kt

```diff
@@ -53,31 +53,64 @@ import androidx.compose.ui.text.input.KeyboardType
 import androidx.compose.ui.unit.dp
 import kotlinx.coroutines.launch
 import org.jellyfin.mobile.R
+import org.jellyfin.mobile.app.ApiClientController
 import org.jellyfin.mobile.setup.ConnectionHelper
 import org.jellyfin.mobile.ui.state.CheckUrlState
-import org.jellyfin.mobile.ui.state.ServerSelectionType
+import org.jellyfin.mobile.ui.state.ServerSelectionMode
 import org.jellyfin.mobile.ui.utils.CenterRow
-import org.jellyfin.sdk.model.api.ServerDiscoveryInfo
-import org.koin.androidx.compose.get
+import org.koin.compose.koinInject
 
 @OptIn(ExperimentalComposeUiApi::class)
+@Suppress("LongMethod")
 @Composable
 fun ServerSelection(
     showExternalConnectionError: Boolean,
-    connectionHelper: ConnectionHelper = get(),
+    apiClientController: ApiClientController = koinInject(),
+    connectionHelper: ConnectionHelper = koinInject(),
     onConnected: suspend (String) -> Unit,
 ) {
     val coroutineScope = rememberCoroutineScope()
     val keyboardController = LocalSoftwareKeyboardController.current
-    var serverSelectionType by remember { mutableStateOf(ServerSelectionType.ADDRESS) }
+    var serverSelectionMode by remember { mutableStateOf(ServerSelectionMode.ADDRESS) }
     var hostname by remember { mutableStateOf("") }
+    val serverSuggestions = remember { mutableStateListOf<ServerSuggestion>() }
     var checkUrlState by remember<MutableState<CheckUrlState>> { mutableStateOf(CheckUrlState.Unchecked) }
     var externalError by remember { mutableStateOf(showExternalConnectionError) }
 
-    val discoveredServers = remember { mutableStateListOf<ServerDiscoveryInfo>() }
+    // Prefill currently selected server if available
+    LaunchedEffect(Unit) {
+        val server = apiClientController.loadSavedServer()
+        if (server != null) {
+            hostname = server.hostname
+        }
+    }
+
+    // Suggest saved servers
+    LaunchedEffect(Unit) {
+        // Append saved servers to list, discovered servers will stay at the front
+        apiClientController.loadPreviouslyUsedServers().mapTo(serverSuggestions) { server ->
+            ServerSuggestion(
+                type = ServerSuggestion.Type.SAVED,
+                name = server.hostname,
+                address = server.hostname,
+                timestamp = server.lastUsedTimestamp,
+            )
+        }
+    }
+
+    // Server discovery
     LaunchedEffect(Unit) {
         connectionHelper.discoverServersAsFlow().collect { serverInfo ->
-            discoveredServers.add(serverInfo)
+            serverSuggestions.removeIf { existing -> existing.address == serverInfo.address }
```

Comment:
This _might_ not always work due to race conditions if the first discovery is made before loadPreviouslyUsedServers returns. I guess it can be easily fixed by putting both in the same LaunchedEffect so discovery runs only after previous used servers are loaded.

Reply:
Yeah, I was ware about that race condition, but couldn't think of a good solution. Just using the same LaunchedEffect just seems so.. easy. I should've thought of that 😅